subsys/llext/llext_load.c

```diff
@@ -372,6 +381,55 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext,
 		size_t top_ofs = MAX(region->sh_offset + region->sh_size,
 				     shdr->sh_offset + shdr->sh_size);
 
+		if (shdr->sh_addralign > region->sh_addralign) {
+			/* This section uses a larger alignment value than what
+			 * is currently used by the region. Make sure its final
+			 * position in memory is correct by adjusting the start
+			 * of the region, if needed.
+			 *
+			 * For example, in such a situation:
+			 *
+			 *   current region: ofs  0x24, size  0xdc, align   0x4
+			 *      new section: ofs 0x100, size  0x20, align  0x10
+			 *
+			 * an aligned allocation would result in the section
+			 * starting at bot_ofs + 0xdc, which would not satisfy
+			 * the original constraint. Rounding the region start
+			 * address by 4 bytes will fix the issue, and this can
+			 * be repeated:
+			 *
+			 *   current region: ofs  0x20, size 0x100, align  0x10
+			 *      new section: ofs 0x200, size  0x20, align 0x100
```

Comment:
Thanks for adding the layout after the adjustment, this is very helpful! Now I can clearly see that there's something that I don't understand: why should the current region's size be extended to 0x100? We moved the beginning down to 0x20, so we add 4 bytes to the size, which is now 0xe0. And why should the new section now begin at 0x200? Why not at 0x100? What I would expect here is that the new region alignment is set to 0x100 and the `.sh_info` records the fact that the region has to start at 0x24 bytes into the allocated buffer. So, you then allocate a 0x100 byte aligned buffer and copy the region at the 0x24 byte offset into it.